src/SFA.DAS.EAS.Portal/Database/Models/Account.cs

```diff
@@ -0,0 +1,116 @@
+
+        [JsonProperty("outboxData")]
+        public IEnumerable<OutboxMessage> OutboxData => _outboxData;
+        
+        [JsonProperty("created")]
+        public DateTime Created { get; private set; }
+
+        [JsonProperty("updated")]
+        public DateTime? Updated { get; private set; }
+
+        [JsonProperty("deleted")]
+        public DateTime? Deleted { get; private set; }
+
+        [JsonIgnore]
+        private readonly List<AccountLegalEntity> _accountLegalEntities = new List<AccountLegalEntity>();
+
+        [JsonIgnore]
+        private List<OutboxMessage> _outboxData = new List<OutboxMessage>();
+
+        private Account(long accountId, DateTime created, string messageId)
+            : base(1)
+        {
+            Id = Guid.NewGuid();
+            AccountId = accountId;
+            // the created date originates from event publishers, so we make sure we store it as UTC
+            Created = DateTime.SpecifyKind(created, DateTimeKind.Utc);
+            
+            AddOutboxMessage(messageId, created);
+        }
+
+        // ctor for when creating account doc from reserved funding
+        // do we accept event, seems a coupling too far!
+        public Account(long accountId, long accountLegalEntityId, string legalEntityName, Guid reservationId,
+            string courseId, string courseName, DateTime startDate, DateTime endDate, DateTime created, string messageId)
+            : this(accountId, created, messageId)
+        {
+            AddReserveFunding(accountLegalEntityId, legalEntityName, reservationId, courseId, courseName, startDate, endDate);
+        }
+
+        [JsonConstructor]
+        private Account()
+        {
+        }
+
+        public void AddReserveFunding(long accountLegalEntityId, string legalEntityName,  Guid reservationId,
+            string courseId, string courseName, DateTime startDate, DateTime endDate, DateTime updated, string messageId)
+        {
+            ProcessMessage(messageId, updated, () =>
+            {
+                var existingAccountLegalEntity = _accountLegalEntities.FirstOrDefault(ale => ale.AccountLegalEntityId == accountLegalEntityId);
+                if (existingAccountLegalEntity == null)
+                    AddReserveFunding(accountLegalEntityId, legalEntityName, reservationId, courseId, courseName, startDate, endDate);
+                else
+                    existingAccountLegalEntity.AddReserveFunding(reservationId, courseId, courseName, startDate, endDate);
+                
+                Updated = updated;
+                Deleted = null;
+            });
+        }
+
+        private void AddReserveFunding(long accountLegalEntityId, string legalEntityName,
+            Guid reservationId, string courseId, string courseName, DateTime startDate, DateTime endDate)
+        {
+            _accountLegalEntities.Add(new AccountLegalEntity(accountLegalEntityId, legalEntityName, reservationId, courseId, courseName, startDate, endDate));
```

Comment:
Doesn't prevent adding a reservation with the same Id twice.

Reply:
Worth noting that this has been raised and fixed as part of this branch, the Cosmos Db has a unique constraint against the AccountId, so the subsequent call to add would fail, retry the message and successfully get the account for the AccountId and add a new reservation.